NDTensors/src/dense/dense.jl

```diff
@@ -3,73 +3,145 @@
+function Dense{DataT}() where {DataT<:AbstractArray}
+  typedDataT = set_eltype_if_unspecified(DataT, default_eltype())
+  ElT = eltype(typedDataT)
+  return Dense{ElT,typedDataT}()
 end
 
-# Construct from a set of indices
-function Dense{ElT,VecT}(inds) where {ElT,VecT<:AbstractArray{ElT}}
-  return Dense(VecT(dim(inds)))
+function Dense{DataT}(inds::Tuple) where {DataT<:AbstractArray}
+  typedDataT = set_eltype_if_unspecified(DataT)
+  ElT = eltype(typedDataT)
+  return Dense{ElT,typedDataT}(inds)
 end
 
-Dense{ElT}(dim::Integer) where {ElT} = Dense(zeros(ElT, dim))
+function Dense{DataT}(::UndefInitializer, inds::Tuple) where {DataT<:AbstractArray}
+  typedDataT = set_eltype_if_unspecified(DataT)
+  ElT = eltype(typedDataT)
+  return Dense{ElT,typedDataT}(undef, inds)
+end
 
-Dense{ElT}(::UndefInitializer, dim::Integer) where {ElT} = Dense(Vector{ElT}(undef, dim))
+function Dense{DataT}(x::Number, inds::Tuple) where {DataT<:AbstractArray}
+  typedDataT = set_eltype_if_unspecified(DataT, typeof(x))
+  ElT = eltype(typedDataT)
+  return Dense{ElT,typedDataT}(fill!(similar(typedDataT, dim(inds)), x))
+end
 
-Dense(::Type{ElT}, dim::Integer) where {ElT} = Dense{ElT}(dim)
+# This function is ill-defined. It cannot transform a complex type to real...
+function Dense{ElR}(data::AbstractArray{ElT}) where {ElR,ElT}
+  return Dense(similartype(typeof(data), ElR)(data))
+end
 
-Dense(x::ElT, dim::Integer) where {ElT<:Number} = Dense(fill(x, dim))
+function Dense{ElT}(inds::Tuple) where {ElT}
+  return Dense{ElT}(dim(inds))
+end
 
-Dense(dim::Integer) = Dense(Float64, dim)
+function Dense{ElT}(dim::Integer) where {ElT<:Number}
+  return Dense{ElT,default_datatype(ElT)}(dim)
+end
 
-Dense(::Type{ElT}, ::UndefInitializer, dim::Integer) where {ElT} = Dense{ElT}(undef, dim)
+Dense{ElT}() where {ElT} = Dense{ElT,default_datatype(ElT)}()
 
-Dense(::UndefInitializer, dim::Integer) = Dense(Float64, undef, dim)
+function Dense(data::DataT) where {DataT<:AbstractArray{<:Any,N}} where {N}
+  if N > 1
+    Dense(vec(data))
+  else
+    Dense{eltype(DataT),DataT}(data)
+  end
+end
 
-Dense{ElT}() where {ElT} = Dense(ElT[])
-Dense(::Type{ElT}) where {ElT} = Dense{ElT}()
+function Dense(DataT::Type{<:AbstractArray{<:Any,N}}, dim::Integer) where {N}
+  if N > 1
+    println("Only vector-based datatypes are currently supported.")
+    throw(TypeError)
+  end
+  DataT = set_eltype_if_unspecified(DataT)
+  ElT = eltype(DataT)
+  return Dense{ElT,DataT}(dim)
+end
 
-setdata(D::Dense, ndata) = Dense(ndata)
+Dense(ElT::Type{<:Number}, dim::Integer) = Dense{ElT}(dim)
 
-#
-# Random constructors
-#
+function Dense(ElT::Type{<:Number}, ::UndefInitializer, dim::Integer)
+  return Dense{ElT,default_datatype(ElT)}(undef, (dim,))
+end
 
-function randn(::Type{StoreT}, dim::Integer) where {StoreT<:Dense}
-  return Dense(randn(eltype(StoreT), dim))
+function Dense(::UndefInitializer, dim::Integer)
+  return Dense{default_eltype(),default_datatype(default_eltype())}(undef, (dim,))
```

Comment:
Maybe a slightly simpler way to write this would be:
```julia
  datatype = default_datatype()
  return Dense{eltype(datatype),datatype}(undef, (dim,))
```